### PR TITLE
[Partitioner] Speed up the update of partition map

### PR DESCRIPTION
**Pull request by ghost**
We can update partition map by iterating users of node but not all of the downstream users of node. The former is faster than the latter which has many duplicate insertion.

### CODE REVIEW

**Comment by lingzhi98** [comment]
@tarun292 Same as https://github.com/pytorch/pytorch/pull/136608, can you help to review this PR? Thanks!

**Review by Skylion007 — DISMISSED** [review]

**Comment by lingzhi98** [comment]
@tarun292 can you take a look? Thanks!

**Comment by carsonwang** [comment]
Adding some background info here. In one of our use cases, we have observed pass_propose_partitions took more than 1 minute. We have proposed totally 4 PRs(#136608, #136616, #136614, #136598) to optimize it and they achieved up to 11x speedup in our test. @tarun292 @SherlockNoMad appreciate if you can take a look if these changes make sense. This will be very useful to unblock our use case.

File: torch/fx/passes/infra/partitioner.py
Comment (tarun292) [inline torch/fx/passes/infra/partitioner.py:170]: So in order for this to work we're assuming that the graph is being traversed in reverse. I think maybe we should add an assert that `target_id is not None` here to ensure that the user node has already been traversed?

Comment (lingzhi98) [reply torch/fx/passes/infra/partitioner.py:170]: Done.

**Comment by lingzhi98** [comment]
@tarun292 any comments? Thanks!

**Comment by lingzhi98** [comment]
@tarun292 Can you approve this PR? Thanks.

**Comment by xwu-intel** [comment]
@ezyang It has been a while. As this is a 9 lines change, could anyone help to review and merge ?

**Review by jgong5 — DISMISSED** [review]

File: torch/fx/passes/infra/partitioner.py
Comment (tarun292) [inline torch/fx/passes/infra/partitioner.py:177]: nit: Change error to `Encountered user node which has not been traversed yet. This should only happen if this is an unsupported node.`

**Comment by tarun292** [comment]
Overall changes lgtm. Please update the error message before merging.

**Review by tarun292 — APPROVED** [review]

**Comment by lingzhi98** [comment]
> Overall changes lgtm. Please update the error message before merging.

Done.

**Review by jgong5 — APPROVED** [review]

**Comment by xwu-intel** [comment]
@jgong5 the CI error seems irrelevant. need to rebase and rerun?

**Comment by lingzhi98** [comment]
> ## Merge failed
> **Reason**: 3 mandatory check(s) failed. The first few are:
> 
> * [BC Lint](https://github.com/pytorch/pytorch/actions/runs/11966749077)
> * [pull](https://github.com/pytorch/pytorch/actions/runs/11966749089)
> * [Lint](https://github.com/pytorch/pytorch/actions/runs/11966749104)
> 
> Dig deeper by [viewing the failures on hud](https://hud.pytorch.org/pytorch/pytorch/commit/841cb18d6471d94d47ad47b1a9c3f6165a6d19c7)
> 
> Details for Dev Infra team

Can anyone merge this PR? I don't know what these failure mean.

File: torch/fx/passes/infra/partitioner.py
Comment (digantdesai) [inline torch/fx/passes/infra/partitioner.py:186]: This is too strict compared to the previous behavior. And causing a failure for previously OK test.

That is the `user_node` here may not be in the `assignment` (perhaps because it is not supported), and we shouldn't assert here.

**Pull request closed without merging**
